lib/jsonapi_parameters/schema_validator.rb

```diff
@@ -0,0 +1,51 @@
+require 'json_schemer'
+
+module JsonApi::Parameters
+  SCHEMA_PATH = Pathname.new(__dir__).join('jsonapi_schema.json').to_s.freeze
+
+  private
+
+  def should_prevalidate?
+    JsonApi::Parameters.enforce_schema_prevalidation && !JsonApi::Parameters.suppress_schema_validation_errors
+  end
+
+  class SchemaValidator
+    class ValidationError < StandardError; end
+
+    def initialize(payload)
+      @payload = payload.deep_stringify_keys
+    end
+
+    attr_reader :payload
+
+    def validate!
+      schema = JSONSchemer.schema(File.read(SCHEMA_PATH))
+
+      unless schema.valid?(payload) # rubocop:disable Style/GuardClause
+        errors = []
+
+        schema.validate(payload).each do |validation_error|
+          errors << nice_error(validation_error)
+        end
```

Comment:
schemer seems not caching validation result so:
```suggestion
      validation_errors = schema.validate(payload)
      unless validation_errors.blank? # rubocop:disable Style/GuardClause
        errors = validation_errors.map do |validation_error|
          nice_error(validation_error)
        end
```